packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -62,6 +63,149 @@ const TextStyle _kToolbarButtonDisabledFontStyle = TextStyle(
 // Eyeballed value.
 const EdgeInsets _kToolbarButtonPadding = EdgeInsets.symmetric(vertical: 10.0, horizontal: 18.0);
 
+// Generates the child that's passed into CupertinoTextSelectionToolbar.
+class _CupertinoTextSelectionToolbarWrapper extends StatefulWidget {
+  const _CupertinoTextSelectionToolbarWrapper({
+    Key key,
+    this.arrowTipX,
+    this.barTopY,
+    this.clipboardStatus,
+    this.handleCut,
+    this.handleCopy,
+    this.handlePaste,
+    this.handleSelectAll,
+    this.isArrowPointingDown,
+  }) : super(key: key);
+
+  final double arrowTipX;
+  final double barTopY;
+  final ClipboardStatusNotifier clipboardStatus;
+  final VoidCallback handleCut;
+  final VoidCallback handleCopy;
+  final VoidCallback handlePaste;
+  final VoidCallback handleSelectAll;
+  final bool isArrowPointingDown;
+
+  @override
+  _CupertinoTextSelectionToolbarWrapperState createState() => _CupertinoTextSelectionToolbarWrapperState();
+}
+
+class _CupertinoTextSelectionToolbarWrapperState extends State<_CupertinoTextSelectionToolbarWrapper> {
+  ClipboardStatusNotifier _clipboardStatus;
+
+  void _onChangedClipboardStatus() {
+    setState(() {
+      // Inform the widget that the value of clipboardStatus has changed.
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _clipboardStatus = widget.clipboardStatus ?? ClipboardStatusNotifier();
+    _clipboardStatus.addListener(_onChangedClipboardStatus);
+    _clipboardStatus.update();
+  }
+
+  @override
+  void didUpdateWidget(_CupertinoTextSelectionToolbarWrapper oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.clipboardStatus == null && widget.clipboardStatus != null) {
+      _clipboardStatus.removeListener(_onChangedClipboardStatus);
+      _clipboardStatus.dispose();
+      _clipboardStatus = widget.clipboardStatus;
+    } else if (oldWidget.clipboardStatus != null) {
+      if (widget.clipboardStatus == null) {
+        _clipboardStatus = ClipboardStatusNotifier();
+        _clipboardStatus.addListener(_onChangedClipboardStatus);
+        oldWidget.clipboardStatus.removeListener(_onChangedClipboardStatus);
+      } else if (widget.clipboardStatus != oldWidget.clipboardStatus) {
+        _clipboardStatus = widget.clipboardStatus;
+        _clipboardStatus.addListener(_onChangedClipboardStatus);
+        oldWidget.clipboardStatus.removeListener(_onChangedClipboardStatus);
+      }
+    }
+    if (widget.handlePaste != null) {
+      _clipboardStatus.update();
+    }
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    if (!_clipboardStatus.disposed) {
```

Comment:
This looks a little fishy, why do you need to check whether it is already disposed?

Reply:
A bunch of tests were failing without this (and the same thing on the material side) with "A ClipboardStatusNotifier was used after being disposed."  Is there a cleaner way to handle this?  I would have thought that the toolbar should be disposed before the EditableText, but that's not what's happening.  Maybe because it's in an overlay? Or the `materialTextSelectionControls` singleton?

Reply:
Is the toolbar getting disposed after the EditableText because the overlay removal may be delayed by one frame:

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/overlay.dart#L131-L142

Reply:
If that's indeed the reason, maybe leave a comment here why this is so strange.

Reply:
That doesn't seem to be the exact reason.  It seems like it is taking the `addPostFrameCallback` path, but the text selection menu is already disposed before the async callback is called.  It must be getting disposed from somewhere else.

I tried to track down where exactly so I could understand this, but I couldn't figure out what is causing it to be disposed and why it happens after EditableText.  Commenting out the code that calls `remove` on the OverlayEntry doesn't even stop it from being disposed.  There must be something I don't understand about how Flutter cleans things like this up internally.

I left a generic comment about why I'm using `disposed`.  If you have more ideas about why then I can look into it and elaborate.

Reply:
To me it looks like this may be the sequence of events:

At the end of the build phase the EditableText is disposed. In its dispose it first calls dispose on the overlay, which ultimately calls the dispose method of the OverlayEntry that I linked above. Since we are still in the build phase (which happens during SchedulerPhase.persistentCallbacks), we can't modify the overlay directly and thus schedule a post frame callback to remove the overlay. Next, the EditableText disposes the clipboard. The build phase ends and the post-frame callbacks run. In one of those callbacks, the overlay entry is finally removed from the overlay causing the dispose callbacks of the toolbar state to execute. At this point the clipboard thing is already disposed and operating on it causes the exceptions you're seeing.

But you're saying the TextSelectionMenu gets disposed before the post framework to remove it from the overlay runs? Not sure I understand that because in your code you're guarding against the clipboard thing being disposed...

Reply:
Here's the order I see by using print statements:

  1. EditableTextState.dispose
  1. ClipboardStatusNotifier.dispose
  1. _TextSelectionToolbarState.dispose
  1. `_overlay.remove(this)` inside of the async callback in OverlayEntry.remove.

So yeah, it doesn't make sense why number 4 is last.